ecs-agent/tmds/server_integ_test.go

```diff
@@ -0,0 +1,187 @@
+func TestRequestWriteTimeout(t *testing.T) {
+	// Setup a simple router with a simple hello world handler and a slow handler
+	router := mux.NewRouter()
+	router.HandleFunc("/", helloWorldHandler())
+	router.HandleFunc("/slow", func(w http.ResponseWriter, r *http.Request) {
+		// add fake delay to the handler
+		time.Sleep(100 * time.Millisecond)
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintln(w, "Hello world")
+	})
+
+	serverAddress := "127.0.0.1:3599"
+
+	// Setup the server with a low write timeout for testing
+	server, err := NewServer(nil,
+		WithRouter(router),
+		WithListenAddress(serverAddress),
+		WithWriteTimeout(50*time.Millisecond),
+		WithSteadyStateRate(10),
+		WithBurstRate(10),
+	)
+	require.NoError(t, err)
+
+	// Start the server
+	startServer(t, server)
+	defer server.Close()
+
+	client := http.DefaultClient
+	err = waitForServer(client, serverAddress)
+	require.NoError(t, err)
+
+	// An EOF error is expected when write timeout is exceeded
+	_, err = client.Get("http://" + serverAddress + "/slow")
+	require.ErrorIs(t, err, io.EOF)
+}
+
+// Tests that the server initialized with NewServer() preserves the routes on the
+// passed router instance.
+func TestRoutes(t *testing.T) {
+	// Setup a simple router with a couple of routes
+	router := mux.NewRouter()
+	router.HandleFunc("/", helloWorldHandler())
+	router.HandleFunc("/products/{id}", func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintf(w, "id:"+mux.Vars(r)["id"])
+	})
+
+	// setup the server
+	serverAddress := "127.0.0.1:3600"
+	server, err := NewServer(nil,
+		WithRouter(router),
+		WithListenAddress(serverAddress),
+		WithSteadyStateRate(10),
+		WithBurstRate(10),
+	)
+	require.NoError(t, err)
+
+	// Start the server
+	startServer(t, server)
+	defer server.Close()
+
+	// wait for the server to come up
+	client := http.DefaultClient
+	err = waitForServer(client, serverAddress)
+	require.NoError(t, err)
+
+	// send a test request to /products/{id} endpoint and read the body
+	res, err := client.Get("http://" + serverAddress + "/products/5")
+	require.NoError(t, err)
+	body, err := ioutil.ReadAll(res.Body)
+	require.NoError(t, err)
+
+	// assertions
+	assert.Equal(t, http.StatusOK, res.StatusCode)
+	assert.Equal(t, "id:5", string(body))
+}
+
+// Returns an HTTP handler that responds with "Hello world"
+func helloWorldHandler() func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+		fmt.Fprintln(w, "Hello world")
+	}
+}
+
+// Waits for the server to come up. Checks if the server is up by sending
+// repeated requests to it.
+func waitForServer(client *http.Client, serverAddress string) error {
+	var err error
+	// wait for the server to come up
+	for i := 0; i < 10; i++ {
+		_, err = client.Get("http://" + serverAddress)
+		if err == nil {
+			return nil // server is up now
+		}
+		time.Sleep(100 * time.Millisecond)
```

Comment:
nit: should this sleep be before Get call?

Reply:
Good point. I will move the sleep to before the HTTP call!

Reply:
Updated in https://github.com/aws/amazon-ecs-agent/pull/3660/commits/d7da34f3a8cbba0b5261e9dc628f54de50936fda